toys/LisExpr.chpl

```diff
@@ -261,32 +325,50 @@ module LisExpr
       tokenize the prog
     */
     proc tokenize(line: string) {
-        return line.replace("("," ( ").replace(")"," ) ").split();
+        // Want:
+        //   var l: list(string) = line.replace("("," ( ").replace(")"," ) ").split()
+        // Workaround (see https://github.com/chapel-lang/chapel/issues/16166):
+
+        var l: list(string);
+        for token in line.replace("("," ( ").replace(")"," ) ").split() do
+          l.append(token);
+        return l;
     }
     
     /*
       parse, check, and validate code and all symbols in the tokenized prog
     */ 
     proc parse(line: string): owned GenListValue throws {
-        return read_from_tokens(tokenize(line));
+        // Want:
+        //   return read_from_tokens(tokenize(line));
+        //
+        // Workaround (see https://github.com/chapel-lang/chapel/issues/16170):
+
+        var l: list(string) = tokenize(line);
+        return read_from_tokens(l);
     }
     
     /*
       parse throught the list of tokens generating the parse tree / AST
       as a list of atoms and lists
     */
-    proc read_from_tokens(tokens: [?D] string): owned GenListValue throws {
+    proc read_from_tokens(ref tokens: list(string)): owned GenListValue throws {
         if (tokens.size == 0) then
             throw new owned ErrorWithMsg("SyntaxError: unexpected EOF");
-        var token = tokens.pop_front();
+
+        // Open Q: If we were to parse from the back of the string to the
+        // front, could this be more efficient since popping from the
+        // front of a list is an expensive operation?
+
+        var token = tokens.pop(0);
         if (token == "(") {
             var L: GenList;
-            while (tokens[D.low] != ")") {
+            while (tokens.first() != ")") {
                 L.append(read_from_tokens(tokens));
                 if (tokens.size == 0) then
                     throw new owned ErrorWithMsg("SyntaxError: unexpected EOF");
             }
-            tokens.pop_front(); // pop off ")"
+            tokens.pop(0); // pop off ")"
             return new owned ListValue(L);
```

Comment:
@mhmerrill: This is the line you're asking about right?  Note that, in this diff, I did not modify it.  That's what makes me say that although it's a potential problem, it's a pre-existing one, not anything specific to what I did in this PR.  Again, unless I'm missing something subtle.

Reply:
Yes, you didn't change this line but you drasticly changed the `ListValue.init` and the `GenListValue.init` making the deep copy. I also think I am gaining an understanding of why you made the change. My question is how to do this and not make the deep copy... The problem is that I create an `owned GenList L` then add a bunch of stuff to it then use it to initialize a `ListValue(L)` that is where all the copying happens. Maybe we should create/initialize the `new owned ListValue(L)` when the list is still empty before we populate it in line 367, and then line 367 might look like `LV.L.append(read_from_tokens(tokens));`
Am I making sense? Also, could we change this to a `shared` to get around this copying problem`